File: LeetCode/Study Plan/Practice Problems/Medium/Binary Tree Zig Zag Order Traversal/Java/binaryTreeZigZag.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;

        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);
        boolean flag = false;

        while(!queue.isEmpty()) {
            int size = queue.size();
            ArrayDeque<TreeNode> currQueue = new ArrayDeque<>();
            List<Integer> currRes = new ArrayList<>();

            for (int i=0; i<size; i++) {
                TreeNode curr = queue.poll();
                if (curr.left != null) currQueue.offer(curr.left);
                if (curr.right != null) currQueue.offer(curr.right);
                currRes.add(curr.val);
            }

            if (flag) Collections.reverse(currRes);
            flag = !flag;
            res.add(currRes);
            queue = currQueue;
        }

        return res;
    }
}

File: LeetCode/Medium/Binary Tree Zig Zag Order Traversal/Go/binaryTreeZigZag.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left,
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val, Right: nil, Left: nil} }

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}
	even := true
	q := []*TreeNode{root}
	res := [][]int{}
	for len(q) > 0 {
		temp := []int{}
		n := len(q)
		for i := 0; i < n; i++ {
			curr := q[0]
			q = q[1:]
			temp = append(temp, curr.Val)
			if curr.Left != nil {
				q = append(q, curr.Left)
			}
			if curr.Right != nil {
				q = append(q, curr.Right)
			}
		}
		if !even {
			reverse(&temp)
		}
		even = !even
		res = append(res, temp)
	}
	return res
}

func reverse(t *[]int) {
	n := len(*t)
	for i := 0; i < n/2; i++ {
		(*t)[i], (*t)[n-i-1] = (*t)[n-1-i], (*t)[i]
	}
}

func main() {
	root := New(3)
	root.Left = New(9)
	root.Right = New(20)
	root.Right.Left = New(15)
	root.Right.Right = New(7)
	fmt.Println(zigzagLevelOrder(root))
}
